### ISSUE TITLE
More basic GDScript methods and maybe iterators?

### ISSUE THREAD

**Issue by hydroper** [godotengine/godot-proposals#7197]
### Describe the project you are working on

RPG

### Describe the problem or limitation you are having in your project

None.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Just simple things taken from JavaScript:

- `array.to_reversed()`: doesn't reverse in-place.
- `array.to_sorted(callback_fn)`: doesn't sort in-place.
- Loop accepts iterables or iterators.
- `dict.entries()` returns an `Iterator[Dictionary]` in the form `{key, value}`
- `dict.keys()` returns an `Iterator[Variant]`
- `dict.values()` returns an `Iterator[Variant]`

Since there's no typing, `Iterable[T]` and `Iterator[T]` are just synonyms of:
- _Iterable:_ value with a `get_iterator()` method.
- _Iterator:_ value with a `next()` method that returns a record `{done: bool, value: T}`.

It's convenient that standard collections be iterables, allowing one to convert to the other through some sort of constructor or conversion function (similiar to `Array.from(iterable)` from JS or `new Map(iterable)`).

### Iterables

An iterator is an iterable. An iterable is an interface that yields an iterator.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

```
for something in self.data.party.to_reversed():
    pass

for entry in dictionary.entries():
    # entry.key
    # entry.value
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Yes, through a namespace, such as `ArrayExt` or `ArrayUtil`.

### Is there a reason why this should be core and not an add-on in the asset library?

Part of the GDScript standard objects.

**Comment by Calinou** [comment]
> An iterator is an iterable. An iterable is an interface that yields an iterator.

Custom iterators already exist in GDScript: https://docs.godotengine.org/en/latest/tutorials/scripting/gdscript/gdscript_advanced.html#custom-iterators

**Comment by hydroper** [comment]
> This could be implemented by making dict.entries() return an array of 2-item arrays, with the first item being the key and the second item being the value. This should result in significantly less overhead than going through the custom iterator system, which requires creating a (heap-allocated) class.

Yes, however the issue is that GDScript has no destructuring, so it will get unreadable to do this:

```
entry[0]
# vs. entry.key
```

Also... about optimization, I think immutable records passed-by-value would solve the issue if GDScript was entirely typed at least like C#, Java or TypeScript.

**Comment by hydroper** [comment]
Oops, I misread it.

**Comment by hydroper** [comment]
Also, on JS, things like `reverse()` return the original array. So if GDScript did the same, you could just do `Array(array).reverse()`, but unfortunately it returns `void`. Anyway, `to_reversed()` is still better.

**Comment by LoipesMas** [comment]
I am glad that GDScript is embracing more functional approach (by adding `Array.map`, `Array.reduce`, etc.), but I feel like it's still lacking in some basic functions, such as aforementioned `Dictionary.entries` (or maybe it should be `Dictionary.items` to retain similarity to Python). 
I would also like to see methods like `Array.min_custom`, `Array.max_custom`, `Array.sum`. All of those can be implemented in GDScript (it's even show in `Array` documentation), but those implementations are quite verbose, so I still feel like they should be in the "standard library", instead of having everyone re-implementing them for themselves. And, I presume, that they could be implemented in C++ instead of GDScript for better performance, but that's a bonus and not my main point.
However, my "requirements" don't involve more iterators, so I'm not sure if I should make a separate proposal for that.

**Comment by MikeSchulze** [comment]
a flatmap would be nice.
`Array.flatmap(accumulator :Callable, mapper :Callable) -> Variant`
This would by allow to map and flating arrrys
e.g.
```
var values :Array[Foo]
var count := values.flatmap(func(acc :Variant, value :Variant): return acc + value, func(foo :Variant): return foo.bars())`
```
